src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java

```diff
@@ -191,18 +218,32 @@ public JsonDeserializer<?> createContextual(DeserializationContext ctxt,
         // 14-Jun-2017, tatu: [databind#1625]: may want to block merging, for root value
         boolean preventMerge = (property == null)
                 && Boolean.FALSE.equals(ctxt.getConfig().getDefaultMergeable(Object.class));
+        // 31-Aug-2024: [databind#4680] Allow custom key deserializer for Object.class deserialization
+        KeyDeserializer keyDeser = ctxt.findKeyDeserializer(SimpleType.constructUnsafe(Object.class), property);
         // 20-Apr-2014, tatu: If nothing custom, let's use "vanilla" instance,
         //     simpler and can avoid some of delegation
         if ((_stringDeserializer == null) && (_numberDeserializer == null)
                 && (_mapDeserializer == null) && (_listDeserializer == null)
                 &&  getClass() == UntypedObjectDeserializer.class) {
-            return UntypedObjectDeserializerNR.instance(preventMerge);
+            if (keyDeser == null) {
+                return UntypedObjectDeserializerNR.instance(preventMerge);
+            }
         }
 
+        UntypedObjectDeserializer untyped  = null;
         if (preventMerge != _nonMerging) {
-            return new UntypedObjectDeserializer(this, preventMerge);
+            untyped = new UntypedObjectDeserializer(this, preventMerge);
+        }
+        if (keyDeser != null) {
+            if (untyped == null) {
+                untyped = new UntypedObjectDeserializer(this, keyDeser);
+            } else {
+                untyped = new UntypedObjectDeserializer(untyped, keyDeser);
+            }
+        }
```

Comment:
@cowtowncoder This is what I came up with atm. I was hoping maybe you might think of better approach than this? 😅

First, null/non-null check for `keyDeser` is done at line 228 and line 237
Second, declaring local variable `untyped` seems a bit unnatural.